src/simulators/matrix_product_state/matrix_product_state_internal.cpp

```diff
@@ -520,14 +520,81 @@ void MPS::apply_matrix(const reg_t & qubits, const cmatrix_t &mat)
   case 1: 
     q_reg_[qubits[0]].apply_matrix(mat);
     break;
-  case 2:
-    apply_2_qubit_gate(qubits[0], qubits[1], su4, mat);
-    break;
+    //  case 2:
+    //    apply_2_qubit_gate(qubits[0], qubits[1], su4, mat);
+    //    break;
   default:
-    throw std::invalid_argument("currently support apply_matrix for 1 or 2 qubits only");
+    apply_multi_qubit_gate(qubits, mat);
   }
 }
 
+void MPS::apply_multi_qubit_gate(const reg_t &qubits,
+				 const cmatrix_t &mat) {
+  uint_t num_qubits = qubits.size();
+  // need to reverse qubits because that is the way they
+  // are defined in the Qiskit interface
+  reg_t reversed_qubits = qubits;
+  std::reverse(reversed_qubits.begin(), reversed_qubits.end()); 
+
+  bool ordered = true;
+  for (uint_t index=0; index < qubits.size()-1; index++) {
+    if (reversed_qubits[index]+1 != reversed_qubits[index+1]){
+      ordered = false;
+      break;
+    }
+  }
+  reg_t actual_indices(num_qubits_);
+  std::iota( std::begin(actual_indices), std::end(actual_indices), 0);
+  reg_t target_qubits(num_qubits);
+  if (ordered) {
+    target_qubits = reversed_qubits;
+  } else {
+    move_qubits_to_right_end(reversed_qubits, target_qubits, actual_indices);
+  }
+
+  uint_t first = target_qubits.front();
+  MPS_Tensor sub_tensor(state_vec_as_MPS(first, first+num_qubits-1));
+
+  sub_tensor.apply_matrix(mat, false /* swapped */);
+
+
+  // state_mat is a matrix containing the flattened representation of the sub-tensor 
+  // into a single matrix. Note that sub_tensor will contain 8 matrices for 3-qubit
+  // gates. state_mat will be the concatenation of them all.
+  cmatrix_t state_mat = sub_tensor.get_data(0);
+  for (uint_t i=1; i<sub_tensor.get_data().size(); i++)
```

Comment:
We are skipping the first element on purpose, right?

Reply:
Is the sum of all `sub_tensor` matrices considered big? Then maybe parallelizing this loop will increase performance:
```c++
#pragma omp for
for(int_t i=1; i < sub_tensor.get_data().size(); ++i) {
...
}
```

Reply:
1. it is okay to skip the first element.
2. The size will depend on the size of the matrix applied. I don't expect it to be large, so not adding parallelization.